### PR TITLE
Tcs34725 automatic sampling settings for improved dynamics and accuracy

### PR DESCRIPTION
**Pull request by swifty99**
# What does this implement/fix?

Feature: add an auto mode for integration time and gain factor. The algorithm tries to achieve high accuracy with good dynamics. The sensor now can deliver proper lux and color temperature from below 1 lux up to 30k lux! 

For my needs I wanted to measure outdoor light with proper color temperature from direct sunshine to almost complete dark. The sensor can deliver, now the software can too.

Further changes:
- In auto mode the readings will never be discarded, even fully saturated. With auto gain this is not very likely to happen often. The old implementation, which is unchanged with fixed timings applied, return a zero brightness after over satuaration which I consider a bad behavior.
- The new default timing is auto, I assume this is what most people would need.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#1951

## Test Environment

- [x] ESP32
- [ ] ESP32 IDF
- [x] ESP8266

## Example entry for `config.yaml`:

```yaml
sensor:
  - platform: tcs34725
    red_channel:
      name: "TCS34725 Red Channel"
    clear_channel:
      name: "TCS34725 Clear Channel"
    illuminance:
      name: "TCS34725 Illuminance"
    color_temperature:
      name: "TCS34725 Color Temperature"
    integration_time: auto
    address: 0x29
    update_interval: 2s
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/tcs34725/tcs34725.cpp [hunk 1]
```diff
@@ -226,14 +229,84 @@ void TCS34725Component::update() {
...
+
+  if (integration_time == 0x100){
+    integration_time_auto_ = true;
+    this->integration_reg_ = TCS34725_INTEGRATION_TIME_154MS;
+    my_integration_time_regval = TCS34725_INTEGRATION_TIME_154MS;
+    }
+  else {
+    this->integration_reg_ = integration_time;
+    my_integration_time_regval = integration_time;
+    integration_time_auto_ = false;
```
Comment (dcousens) [inline esphome/components/tcs34725/tcs34725.cpp]:
```suggestion
    this->integration_time_auto_ = false;
```

See https://esphome.io/guides/contributing.html#codebase-standards

File: esphome/components/tcs34725/tcs34725.cpp [hunk 2]
```diff
@@ -226,14 +229,84 @@ void TCS34725Component::update() {
...
   this->status_clear_warning();
 }
 void TCS34725Component::set_integration_time(TCS34725IntegrationTime integration_time) {
-  this->integration_reg_ = integration_time;
-  this->integration_time_ = (256.f - integration_time) * 2.4f;
+  // if an integration time is 0x100, this is auto start with 154ms as this gives best starting point
+  TCS34725IntegrationTime my_integration_time_regval;
+
+  if (integration_time == 0x100){
+    integration_time_auto_ = true;
```
Comment (dcousens) [inline esphome/components/tcs34725/tcs34725.cpp]:
```suggestion
    this->integration_time_auto_ = true;
```

File: esphome/components/tcs34725/tcs34725.cpp [hunk 3]
```diff
@@ -149,14 +148,18 @@ void TCS34725Component::calculate_temperature_and_lux_(uint16_t r, uint16_t g, u
   g2 = g - ir;
   b2 = b - ir;
 
+  // discarding super low values? not recemmonded, and avoided by using auto gain.
   if (r2 == 0) {
-    return;
+    ESP_LOGW(TAG, "No light detected on red channel, switch to auto gain or adjust timing, values will be unreliable");
+    // legacy code
+    if (!integration_time_auto_){
```
Comment (dcousens) [inline esphome/components/tcs34725/tcs34725.cpp]:
```suggestion
    if (!this->integration_time_auto_) {
```

**Review by dcousens — COMMENTED** [review]
A few code standard nits, see https://esphome.io/guides/contributing.html#codebase-standards

**Comment by swifty99** [comment]
@[dcousens](https://github.com/dcousens) thx for the comments, I will review and update tomorrow.

File: esphome/components/tcs34725/tcs34725.cpp (hunk 3 above)
Comment (swifty99) [reply esphome/components/tcs34725/tcs34725.cpp]: fixed in https://github.com/esphome/esphome/pull/3258/commits/1972d5e78f768fc501c80b33c09e107b8739a577

**Comment by swifty99** [comment]
oh, boy, I need a better clang workflow.

**Comment by swifty99** [comment]
Hi everyone.

Could someone please have a look at this PR?
It works stable since weeks now, I do not see any reason to delay merging. The old implementation does not let the sensor show its great capabilities.

Jan

File: esphome/components/tcs34725/tcs34725.cpp [hunk 4]
```diff
@@ -220,20 +232,95 @@ void TCS34725Component::update() {
...
+  }
   this->status_clear_warning();
 }
 void TCS34725Component::set_integration_time(TCS34725IntegrationTime integration_time) {
-  this->integration_reg_ = integration_time;
-  this->integration_time_ = (256.f - integration_time) * 2.4f;
+  // if an integration time is 0x100, this is auto start with 154ms as this gives best starting point
+  TCS34725IntegrationTime my_integration_time_regval;
+
+  if (integration_time == 0x100) {
```
Comment (jesserockz) [inline esphome/components/tcs34725/tcs34725.cpp]:
```suggestion
  if (integration_time == TCS34725_INTEGRATION_TIME_AUTO) {
```

File: esphome/components/tcs34725/tcs34725.h [hunk 5]
```diff
@@ -77,10 +78,11 @@ class TCS34725Component : public PollingComponent, public i2c::I2CDevice {
   float glass_attenuation_{1.0};
   float illuminance_;
   float color_temperature_;
+  bool integration_time_auto_{true};
 
  private:
   void calculate_temperature_and_lux_(uint16_t r, uint16_t g, uint16_t b, uint16_t c);
-  uint8_t integration_reg_{TCS34725_INTEGRATION_TIME_2_4MS};
+  uint16_t integration_reg_{TCS34725_INTEGRATION_TIME_AUTO};
```
Comment (jesserockz) [inline esphome/components/tcs34725/tcs34725.h]:
```suggestion
  uint16_t integration_reg_;
```

**Review by jesserockz — CHANGES_REQUESTED** [review]

**Review by jesserockz — APPROVED** [review]

**Pull request merged**